src/Utilities/CleverDicts.jl

```diff
@@ -46,11 +54,28 @@ index_to_key(::Type{MyKey}, i::Int) = MyKey(i)
 key_to_index(key::MyKey) = key.x
 ```
 """
-mutable struct CleverDict{K, V} <: AbstractDict{K, V}
+mutable struct CleverDict{K, V, F, I} <: AbstractDict{K, V}
```

Comment:
`F<:Function`

Reply:
Is this going to break the downstream solvers?

Reply:
Haven't tested, but it is possible.
After writing the entire thing I think that it ok to use the function overloading for now.

Reply:
Yet another non-breaking possibility is to have AbstractCleverDict with 4 params and the Clever Dict with 2, no super good though...

Reply:
What's the hit of just making them `::Function`? Or making `index_to_key` dependent on the `K-V` pair: `key_to_index(::CleverDict{K, V}, x::K)::Int` and `index_to_key(::CleverDict{K, V}, v::Int)::K`,

Reply:
just `::Function` is not good for inference, and the type wont be concrete.

I like:
> making index_to_key dependent on the K-V pair: key_to_index(::CleverDict{K, V}, x::K)::Int and index_to_key(::CleverDict{K, V}, v::Int)::K

Reply:
If you go the second route, we will need a deprecated fallback.